aten/src/ATen/native/TensorShape.cpp

```diff
@@ -61,61 +61,111 @@ static bool sizes_match_except(IntList s1, IntList s2, int64_t dim_except /* sho
+      // don't waste time doing this operation unless i > 0.
+      if (i > 0) {
+        idxs[wrapped].narrow(0, col, this_piece_size) += cumulative_offset;
+      }
+      cumulative_offset += t.size(wrapped);
+      col += this_piece_size;
+    }
+    auto sizes_copy = sizes.vec();
+    sizes_copy[wrapped] = cumulative_offset;
+    return native::sparse_coo_tensor(idxs, vals, sizes_copy, tensors[0].options());
+  }
+  else {
+    // Catting along a dense dimension requires us to create new values.
+    // For illustration, consider the sparse 3d tensors t1 and t2,
+    // given by t1 = [[[1,2],[3,4]], ... (zeros) ..., [[5,6],[7,8]]]
+    // and t2 = [... (zeros) ..., [[9, 10], [11,12]], ... (zeros) ...],
+    // Their concatenation along dimension 2 is:
+    // [[[1,2,0,0],[3,4,0,0]], ... (zeros) ..., [[0,0,9,10],[0,0,11,12]], ... (zeros) ..., [[5,6,0,0],[7,8,0,0]]]
+    //
+    // Their values tensors are, respectively,
+    // [[[1,2],[3,4]],[[5,6],[7,8]]] and [[[9,10],[11,12]]].
+    //
+    // and so the values tensor of their concatenation along dim 2 will be:
+    // [[[1,2,0,0],[3,4,0,0]],[[5,6,0,0],[7,8,0,0]],[[0,0,9,10],[0,0,11,12]]]
+    //
+    // which we can get by taking the values tensor of each tensor, catting it with zeros of the appropriate size on the left and right,
+    // and then catting all those results together.
+
+    // The dimension in each tensor's values object that corresponds to the overall dimension along which we're catting.
+    int64_t values_dim = wrapped - sparse_dim + 1;
+    // The final size along the catted dimension.
+    int64_t total_size = std::accumulate(tensors.begin(), tensors.end(), 0, [values_dim](int64_t l, Tensor const &r) {
+      return l + r._values().size(values_dim);
+    });
+    auto zeros_sizes = tensors[0]._values().sizes().vec();
+    int64_t cumulative_size = 0;
+    std::vector<Tensor> vals_pieces;
+    std::vector<Tensor> idxs_pieces;
+    for (size_t i = 0; i < tensors.size(); ++i) {
+      auto const &t = tensors[i];
+      check_cat_sparse_dims(t, i, sizes, dim, wrapped, sparse_dim, dense_dim);
+      // dimension 0 of values corresponds to the number of values,
+      // rather than to any logical dimension of the sparse tensor.
+      zeros_sizes[0] = t._values().size(0);
+      zeros_sizes[values_dim] = cumulative_size;
+      cumulative_size += t._values().size(values_dim);
+      auto z1 = native::zeros(zeros_sizes, t._values().options());
+      zeros_sizes[values_dim] = total_size - cumulative_size;
+      auto z2 = native::zeros(zeros_sizes, t._values().options());
+      vals_pieces.push_back(native::cat({z1, t._values(), z2}, values_dim));
+      idxs_pieces.push_back(t._indices());
     }
-    cumulative_offset += t.size(wrapped);
-    col += this_piece_size;
+    auto sizes_copy = sizes.vec();
+    sizes_copy[wrapped] = total_size;
+    return native::sparse_coo_tensor(native::cat(idxs_pieces, 1), native::cat(vals_pieces), sizes_copy, tensors[0].options());
```

Comment:
does this mean catting a dense dim results to a uncoalesced sparse tensor? Should we add a note to that?